app/models/flag.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+class Flag < ActiveRecord::Base
+  scope :enabled, -> { where(enabled: true) }
+  scope :system, -> { where("id < 1000") }
+
+  before_save :set_position
+  before_save :set_name_key
+  after_save :reset_flag_settings!
+  after_destroy :reset_flag_settings!
+
+  def used?
+    PostAction.exists?(post_action_type_id: self.id) ||
+      ReviewableScore.exists?(reviewable_score_type: self.id)
+  end
+
+  def self.reset_flag_settings!
+    PostActionType.reload_types
+    ReviewableScore.reload_types
+  end
+
+  def system?
+    self.id < 1000
+  end
+
+  def applies_to?(type)
+    self.applies_to.include?(type)
+  end
+
+  private
+
+  def reset_flag_settings!
+    self.class.reset_flag_settings!
+  end
+
+  def set_position
+    self.position = Flag.maximum(:position).to_i + 1 if !self.position
+  end
+
+  def set_name_key
+    self.name_key = self.name.gsub(" ", "_").gsub(/[^\w]/, "").downcase
```

Comment:
Might be worth adding a couple of unit tests for the model with variations on the name to make sure that this name_key method works OK (e.g. with special characters, things like `It's Illegal`)